### PR TITLE
[SYCL] Add marray support to integer built-in functions

### PR DESCRIPTION
**Pull request by dm-vodopyanov**
This patch adds support of sycl::marray to integer built-in functions (SYCL 2020 spec, Table 178).

### CODE REVIEW

File: sycl/include/sycl/builtins.hpp [hunk 1]
```diff
@@ -1310,6 +1311,260 @@ mul24(T x, T y) __NOEXC {
...
+__SYCL_MARRAY_INTEGER_FUNCTION_BINOP_U_OVERLOAD(mul_hi, x, y, x[j], y[j])
+__SYCL_MARRAY_INTEGER_FUNCTION_BINOP_I_OVERLOAD(mul_hi, x, y, x[j], y[j])
+__SYCL_MARRAY_INTEGER_FUNCTION_BINOP_U_OVERLOAD(rotate, v, i, v[j], i[j])
+__SYCL_MARRAY_INTEGER_FUNCTION_BINOP_I_OVERLOAD(rotate, v, i, v[j], i[j])
+__SYCL_MARRAY_INTEGER_FUNCTION_BINOP_U_OVERLOAD(sub_sat, x, y, x[j], y[j])
+__SYCL_MARRAY_INTEGER_FUNCTION_BINOP_I_OVERLOAD(sub_sat, x, y, x[j], y[j])
+
+#undef __SYCL_MARRAY_INTEGER_FUNCTION_BINOP_I_OVERLOAD
+#undef __SYCL_MARRAY_INTEGER_FUNCTION_BINOP_I_RET_U_OVERLOAD
+#undef __SYCL_MARRAY_INTEGER_FUNCTION_BINOP_U_OVERLOAD
```
Comment (AlexeySachkov) [inline sycl/include/sycl/builtins.hpp:1429]:
```suggestion
#undef __SYCL_MARRAY_INTEGER_FUNCTION_BINOP_U_OVERLOAD
#undef __SYCL_MARRAY_INTEGER_FUNCTION_BINOP_U_2ND_ARG_SCALAR_OVERLOAD
#undef __SYCL_MARRAY_INTEGER_FUNCTION_BINOP_I_2ND_ARG_SCALAR_OVERLOAD
```

File: sycl/include/sycl/builtins.hpp [hunk 2]
```diff
@@ -1310,6 +1311,260 @@ mul24(T x, T y) __NOEXC {
...
+                                                c[j])
+__SYCL_MARRAY_INTEGER_FUNCTION_TEROP_I_OVERLOAD(mad_hi, a, b, c, a[j], b[j],
+                                                c[j])
+__SYCL_MARRAY_INTEGER_FUNCTION_TEROP_U_OVERLOAD(mad_sat, a, b, c, a[j], b[j],
+                                                c[j])
+__SYCL_MARRAY_INTEGER_FUNCTION_TEROP_I_OVERLOAD(mad_sat, a, b, c, a[j], b[j],
+                                                c[j])
+
+#undef __SYCL_MARRAY_INTEGER_FUNCTION_TEROP_I_OVERLOAD
+#undef __SYCL_MARRAY_INTEGER_FUNCTION_TEROP_U_OVERLOAD
```
Comment (AlexeySachkov) [inline sycl/include/sycl/builtins.hpp:1481]:
```suggestion
#undef __SYCL_MARRAY_INTEGER_FUNCTION_TEROP_U_OVERLOAD
#undef __SYCL_MARRAY_INTEGER_FUNCTION_TEROP_U_2ND_3RD_ARGS_SCALAR_OVERLOAD
#undef __SYCL_MARRAY_INTEGER_FUNCTION_TEROP_I_2ND_3RD_ARGS_SCALAR_OVERLOAD
```

**Review by AlexeySachkov — APPROVED** [review]

**Comment by dm-vodopyanov** [comment]
Unrelated to this patch:

Failed Tests (2):
  SYCL :: AtomicRef/atomic_memory_order_acq_rel.cpp
  SYCL :: GroupAlgorithm/reduce_sycl2020.cpp

**Review by steffenlarsen — APPROVED** [review]
Looks good! I wonder if we could eventually consolidate some of the macros, but I think that would be a bigger refactoring.

**Pull request merged**